### PR TITLE
Fix `DragValue` editing

### LINKED ISSUE #2818: `DragValue` edit is discarded when clicked off
**Describe the bug**
If I click a `DragValue` and type in a new value then that new value is discarded if I click on another `DragValue` or `TextEdit`.

Video: https://youtube.com/shorts/_wxqB0XjkIY

### PR DESCRIPTION
**Pull request by Barugon**
Closes #2818.

Use `Response::changed` instead of `Response::lost_focus` (which is problematic).

### CODE REVIEW

**Review by paul-hansen — APPROVED** [review]
This does mean that it will change the value as you are typing now, which essentially reverts what #2687 wanted. Pinging @harrisonmg to give them a chance to comment. 

I would prefer the way this PR has it anyway. Updating the value on every keypress allows you to see the change happening in real time as you type for quickly previewing small adjustments, otherwise you have to press enter and re-focus it every time you want to preview the change. Think about positioning objects precisely in a 3d scene. So I think this is better anyway.

You can always make your own temporary variable and only update it when focus is lost to recreate the functionality #2687 wanted:
```rust
// see the hello_world example for the where self.age would come from
let mut temp_value =
    ui.data_mut(|data| *data.get_temp_mut_or(Id::new("my_temp"), self.age));
let response = ui.add(egui::DragValue::new(&mut temp_value));
if response.changed() {
    ui.data_mut(|data| data.insert_temp(Id::new("my_temp"), temp_value));
}
if response.lost_focus() {
    self.age = temp_value;
}
```
Obviously would want to wrap that in a function or your own widget if you use it a lot.

tldr; I approve of this PR's change, but it should be noted that it is changing the functionality back closer to what it was, but I think it's for the better anyway.

**Comment by Barugon** [comment]
> This does mean that it will change the value as you are typing now

Yes, it's kind of a middle ground. Previously, it would parse the text every frame. With this change, it would only do that work when the `TextEdit` changes. After looking extensively at the focus logic, it's not possible to use `lost_focus` as there are too many issues with it. The focus logic needs a lot of work to make it reliable enough before using `lost_focus`.

**Comment by Barugon** [comment]
I prefer it updating while editing the value also.

**Comment by harrisonmg** [comment]
In my case, I'm updating motor controller tunings, so this behavior presents a bit of an issue for that.

Seems like an option might be nice in the future.

**Comment by paul-hansen** [comment]
It sounds like the problem you want to solve is that updating your motor controller tunings too frequently causes problems with the motor. Personally I wouldn't see that as a problem that a UI library should be making changes to try to solve. Instead, you should use something like a debounce function when setting the motor's values from anywhere to prevent it from being updated more often than it can support. This would be more robust and would prevent things like users still being able to focus and unfocus inputs more quickly than is supported.

That said, if there's a way to add an option to update the value only on focused lost I wouldn't be opposed to it. It needs to work correctly when tabbing and unfocusing by clicking on other dragvalues/widgets, which is hard to do because of how DragValue changes it's widget based on if it's focused or not, which causes checks using lost_focus in DragValue's internal logic to be inconsistent.

**Comment by Barugon** [comment]
> In my case, I'm updating motor controller tunings, so this behavior presents a bit of an issue for that.
> 
> Seems like an option might be nice in the future.

Is it possible, in your code, to update on `lost_focus`?

**Comment by harrisonmg** [comment]
> > In my case, I'm updating motor controller tunings, so this behavior presents a bit of an issue for that.
> > 
> > Seems like an option might be nice in the future.
> 
> Is it possible, in your code, to update on `lost_focus`?

Maybe, I'll probably stick to an older egui version for now.

Specifically, my issue is that typing "100" results in the gain being set to 1, then 10, then 100. It's the order of magnitude changes, not the frequency.

Anyway, I do support this change especially if it improves consistency.

**Comment by emilk** [comment]
I think we should add an option for this, so users can control what behavior they want. I definitely see the use in both

**Comment by harrisonmg** [comment]
> it's not possible to use lost_focus as there are too many issues with it

@Barugon could you give a few examples of when you found that this fails? I think it would be helpful to drive future improvements to the focus logic if need be.

**Comment by Barugon** [comment]
> > it's not possible to use lost_focus as there are too many issues with it
> 
> @Barugon could you give a few examples of when you found that this fails? I think it would be helpful to drive future improvements to the focus logic if need be.

Edit the value and then click on another `DragValue` or press tab and your edits are lost.

**Comment by Barugon** [comment]
There's a video in the issue linked above that show the issue.

**Comment by harrisonmg** [comment]
Ok, so nothing beyond what's described in #2142?

If so, seems like this option + fixing 2142 would be perfect for everyone

**Comment by Barugon** [comment]
> Ok, so nothing beyond what's described in #2142?
> 
> If so, seems like this option + fixing 2142 would be perfect for everyone

Loosing your edit due to tabbing or clicking on another `DragValue` (or `TextEdit`) are actually two separate issues that cause the same problem.

**Comment by Barugon** [comment]
In #2818 I found the culprit for clicking off but it doesn't fix tabbing.

**Comment by tajtiattila** [comment]
@Barugon This seems to fix the main issue in #2877, thank you!

**Comment by Barugon** [comment]
> @Barugon This seems to fix the main issue in #2877, thank you!

Unfortunately hasn't been merged though.

**Comment by Trivaxy** [comment]
Would be nice to see this merged. Some of my users complained that the way it is now is sort of unintuitive and annoying, especially since we're doing rapid data entry on many different `DragValue`s where it's a cycle of "type into box -> click on the next one -> repeat"

**Comment by Barugon** [comment]
@emilk 
Are there more changes you require?

File: crates/egui/src/widgets/drag_value.rs [hunk 1]
```diff
@@ -352,6 +354,12 @@ impl<'a> DragValue<'a> {
         }
         .custom_parser(|s| i64::from_str_radix(s, 16).map(|n| n as f64).ok())
     }
+
+    /// Update the value when there are changes while editing. (Default: true)
```
Comment (emilk) [inline crates/egui/src/widgets/drag_value.rs:358]:
```suggestion
    /// Update the value on each key press when text-editing the value.
    ///
    /// Default: `true`.
    /// If `false`, the value will only be updated when user presses enter or deselects the value.
```

**Review by emilk — APPROVED** [review]

**Comment by emilk** [comment]
Thanks!
